pkg/image/controller/controller.go

```diff
@@ -29,6 +123,73 @@ func needsImport(stream *api.ImageStream) bool {
 	return stream.Annotations == nil || len(stream.Annotations[api.DockerImageRepositoryCheckAnnotation]) == 0
 }
 
+func (c *ImportController) handleImport() {
+	for {
+		select {
+		case <-c.stopChan:
+			return
+
+		case staleImageStream := <-c.work:
+			glog.V(1).Infof("popped %s from the worklist", workingSetKey(staleImageStream))
+
+			// if we're already in the workingset, that means that some thread is already trying to do an import for this.
+			// This does NOT mean that we shouldn't attempt to do this work, only that we shouldn't attempt to do it now.
+			if c.isInWorkingSet(staleImageStream) {
```

Comment:
> isInWorkingSet and addToWorkingSet should be a single call addToWorkingSet() (added bool), otherwise two threads can both work on the same one at the same time

how embarrassing.